hnix-store-core/src/System/Nix/Hash.hs

```diff
@@ -0,0 +1,105 @@
+{-|
+Description : Nix-style hashes (truncated sha256)
+Maintainer  : Shea Levy <shea@shealevy.com>
+-}
+
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+module System.Nix.Hash (
+
+  -- * Introduce hashes for the store
+    hash
+  , hashlazy
+  , fromBase32
+
+  -- * cryptohash-sha256 style incremental hash building
+  , init
+  , update
+  , finalize
+
+  -- * Internal
+  , StorePathHash (..)
+  , truncate52
+  , toNixBase32
+
+  ) where
+
+import qualified Crypto.Hash.SHA256           as SHA
+import           Data.Bits
+import qualified Data.ByteString              as BS
+import qualified Data.ByteString.Lazy         as BSL
+import qualified Data.ByteString.Lazy.Builder as BSL
+import           Data.Char
+import qualified Data.Hashable                as Hashable
+import           Data.Semigroup               ((<>))
+import           Data.Word
+import qualified Data.Vector.Unboxed          as UV
+import           Prelude                      hiding (init)
+
+-- | A string, file, or NAR hash in the format
+--   used in prefixing files in the nix store
+newtype StorePathHash =
+  StorePathHash { getTruncatedHash :: BS.ByteString }
+  deriving (Eq, Hashable.Hashable, Ord, Show)
+
+
+init :: SHA.Ctx
+init = SHA.init
+
+
+update :: SHA.Ctx -> BS.ByteString -> SHA.Ctx
+update = SHA.update
+
+
+finalize :: SHA.Ctx -> StorePathHash
+finalize ctx = StorePathHash . truncate52 $ SHA.finalize ctx
+
+
+hash :: BS.ByteString -> StorePathHash
+hash bs = StorePathHash . BSL.toStrict . toNixBase32 . BSL.fromStrict . truncate' $ SHA.hash bs
+
+
+hashlazy :: BSL.ByteString -> StorePathHash
+hashlazy bs = StorePathHash . BSL.toStrict . toNixBase32 . BSL.fromStrict . truncate' $ SHA.hashlazy bs
+
+
+-- | Import and validate a store path hash
+fromBase32 :: BS.ByteString -> Maybe StorePathHash
```

Comment:
base 32/base 16 questions should just be about presentation, not about internal represenation